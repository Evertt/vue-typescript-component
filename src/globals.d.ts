/**
 * global type declarations in this project
 * should not expose to userland
 */

declare const process: {
	env: {
		NODE_ENV: string,
	},
}

declare interface Array<T> {
	mapToObject(transform: (obj: any) => any): any
}
