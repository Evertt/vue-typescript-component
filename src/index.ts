export * from './prop'
export * from './watch'
export * from './lifecycle'
export * from './component'
