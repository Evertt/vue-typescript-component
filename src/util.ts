/// <reference path="./globals.d.ts" />

export function objAssign<T, U, V>(target: T, source1: U, source2?: V): T & U & V {
	if (source1) {
		for (let n of Object.getOwnPropertyNames(source1)) {
			Object.defineProperty(target, n, Object.getOwnPropertyDescriptor(source1, n))
		}
	}
	if (source2) {
		for (let n of Object.getOwnPropertyNames(source2)) {
			Object.defineProperty(target, n, Object.getOwnPropertyDescriptor(source2, n))
		}
	}
	return <T & U & V>target
}

export function warn (message: string): void {
	if (process.env.NODE_ENV !== 'production' && typeof console !== 'undefined') {
		console.warn('[vue-typescript-component] ' + message)
	}
}
