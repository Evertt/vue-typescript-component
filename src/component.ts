import Vue = require('vue')
import { VueClass } from './vue-class'
import { objAssign } from './util'
import { decorators } from './decorators'

export type ComponentOptions = Vue.ComponentOptions<Vue>

Array.prototype.mapToObject =
	function(transform: (obj: any) => any): any {
		let transformed: any = {}
		for (let elem of this) {
			objAssign(transformed, transform(elem))
		}
		return transformed
	}

export interface NoArgumentConstructable {
	new (): any
	name?: any
	vueComponentOptions?: ComponentOptions
}

function isNoArgumentConstructable(arg: any): arg is NoArgumentConstructable {
	return arg instanceof Function
}

let makeObject = (obj: any) => (key: string) => ({ [key]: obj[key] })
let getMethod  = (obj: any) => (key: string) => Object.getOwnPropertyDescriptor(obj, key).value

function makeComponentDecorator(options: ComponentOptions = {}): ComponentDecorator {
	return (Component: NoArgumentConstructable) => {
		const component = new Component()
		const prototype = Component.prototype

		const superProto = Object.getPrototypeOf(prototype)
		const Super = superProto instanceof Vue
			? superProto.constructor as VueClass
			: Vue

		let andDecorate = (obj: any) => (key: string) =>
			key == 'constructor' || ! (decorators[key] || []).reduce(
				(result, decorator) =>
					decorator.target == obj || decorator.target == obj.__proto__
						? decorator.fn(options, obj) || result : result, false,
			)

		let componentKeys = Object.getOwnPropertyNames(component).filter(andDecorate(component))
		let prototypeKeys = Object.getOwnPropertyNames(prototype).filter(andDecorate(prototype))

		let decKeys = Object.keys(decorators)
		let allKeys = componentKeys.concat(prototypeKeys)
		decKeys.filter(key => allKeys.indexOf(key) === -1).filter(andDecorate(component))

		let staticKeys = Object.getOwnPropertyNames(Component)
		let superStaticKeys = Object.getOwnPropertyNames(Super)
		let componentStaticKeys = staticKeys.filter( key => superStaticKeys.indexOf(key) === -1)

		componentKeys = componentKeys.filter(andDecorate(component))
		prototypeKeys = prototypeKeys.filter(andDecorate(prototype))

		getData(componentKeys, options, component)
		getMethods(prototypeKeys, options, prototype)
		getComputed(prototypeKeys, options, prototype)
		getOptions(componentStaticKeys, options, Component)

		const superOptions = Component.vueComponentOptions || {}
		options.name       = options.name || (<any>Component).name
		options.methods    = objAssign( {}, superOptions.methods,  options.methods  )
		options.computed   = objAssign( {}, superOptions.computed, options.computed )
		options.watch      = objAssign( {}, superOptions.watch,    options.watch    )
		options.props      = objAssign( {}, superOptions.props,    options.props    )

		Component.vueComponentOptions = options

		if (process.env.NODE_ENV !== 'test') return Super.extend(options)
	}
}

function getData(keys: string[], options: ComponentOptions, component: any) {
	let dataKeys = keys.filter(key => key[0] !== '_' && key[0] !== '$')
	options.data = () => dataKeys.mapToObject(makeObject(component))
}

function getMethods(keys: string[], options: ComponentOptions, prototype: any) {
	let dataKeys = keys.filter(key => key !== 'constructor')

	let methods = (options.methods || (options.methods = {}))

	for (let key of dataKeys) {
		let method = getMethod(prototype)(key)
		if (method) { methods[key] = method }
	}
}

function getComputed(keys: string[], options: ComponentOptions, prototype: any) {
	let computedKeys = keys.filter(key => Object.getOwnPropertyDescriptor(prototype, key).get)

	options.computed = computedKeys.mapToObject( key => {
		let pd = Object.getOwnPropertyDescriptor(prototype, key)
		return { [key]: { get: pd.get, set: pd.set } }
	})
}

function getOptions(keys: string[], options: ComponentOptions, Component: any) {
	objAssign(options, keys.mapToObject(makeObject(Component)))
}

/** Create property constructor.vueComponentOptions based on method/field annotations
 *  If provided, use options as the base value (.data is always overridden)
 */
export type ComponentDecorator = (cls: NoArgumentConstructable) => void
export function Component(cls: NoArgumentConstructable): void
export function Component(options: ComponentOptions): ComponentDecorator
export function Component(options: ComponentOptions | NoArgumentConstructable): ComponentDecorator | void {
	if (isNoArgumentConstructable(options)) {
		return makeComponentDecorator()(options)
	} else {
		return makeComponentDecorator(options)
	}
}

export const component = Component
