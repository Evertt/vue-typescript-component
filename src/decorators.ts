import { ComponentOptions } from './component'

export type Decorator = (options: ComponentOptions) => void | boolean

export let decorators: { [key: string]: Decorator[] } = {}

export function addDecorator(key: string, decorator: Decorator) {
	let list = decorators[key] || []
	list.push(decorator)
	decorators[key] = list
}
