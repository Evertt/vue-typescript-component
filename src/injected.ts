import { addDecorator } from './decorators'

export function injected(target: any, key: string): void {
    addDecorator(key, () => true)
}

export const Injected = injected
