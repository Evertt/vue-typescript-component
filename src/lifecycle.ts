import { warn } from './util'
import { addDecorator } from './decorators'

const lifeCycleHooks = [
		'beforeCreate', 'created',
		'beforeMount', 'mounted',
		'activated', 'deactivated',
		'beforeUpdate', 'updated',
		'beforeDestroy', 'destroyed',
		'render', /* not a lifecyle hook, but handled identically */
]

export function lifecycle(target: any, key: string): void {
	if (lifeCycleHooks.indexOf(key) === -1) {
		return warn(
			'The @lifecycle decorator only ' +
			'applies to valid lifecycle hooks ' +
			'and the render method.',
		)
	}

	addDecorator(key, options => {
		(options as any)[key] = Object.getOwnPropertyDescriptor(target, key).value

		return true
	})
}

export const Lifecycle = lifecycle
