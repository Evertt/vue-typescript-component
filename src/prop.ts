import Vue = require('vue')
import { addDecorator } from './decorators'

export function prop(target: any, key: string): void {
	addDecorator(key, options => {
		let property: Vue.PropOptions = {}

		if (typeof target[key] === 'object') {
			property.default = () => target[key]
		} else {
			property.default = target[key]
		}

		if (property.type === undefined) {
			try {
				property.type = Object.getPrototypeOf(target[key]).constructor
			} catch (error) {
				// property.type = target[key].__proto__ ? target[key].__proto__.constructor : target[key].constructor
			}
		}

		if (property.default === undefined) {
			property.required = true
		}

		(options.props || (options.props = {}) as any)[key] = property
	})
}

export const Prop = prop
