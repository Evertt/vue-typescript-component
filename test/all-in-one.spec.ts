/// <reference path='../node_modules/@types/jest/index.d.ts' />

import Vue = require('vue')
import { Component, prop, watch, lifecycle } from '../src/index'

@Component
class AllInOne extends Vue {
	aString = 'abc'
	aNumber = 123
	// tslint:disable-next-line:variable-name
	_thisWillBeIgnored = 345
	$asWillThis = 345

	@prop aStringPropWithValue = 'abc'
	@prop aNumberPropWithValue = 123

	@prop aStringProp: string
	@prop aNumberProp: number

	@prop(String) anExplicitStringProp: string
	@prop([String, Number]) anExplicitlyTypedProp: string | number

	get aComputedString(): string { return this.aString }
	set aComputedString(value: string) { this.aString = value }

	get aComputedNumber(): number { return this.aNumber }
	set aComputedNumber(value: number) { this.aNumber = value }

	get aComputedStringGetter(): string { return this.aString }
	get aComputedNumberGetter(): number { return this.aNumber }

	aMethod() { return 'aMethod' }

	// a lifecycle hook
	@lifecycle created() { /* do nothing */ }

	@watch('aString') aStringWatch(val: string, oldVal: string) { /* do nothing */ }
}

it('creates the expected options', () => {
	expect((<any>AllInOne).vueComponentOptions).toMatchSnapshot()
})
it('creates the expected data', () => {
	expect((<any>AllInOne).vueComponentOptions.data()).toMatchSnapshot()
})
