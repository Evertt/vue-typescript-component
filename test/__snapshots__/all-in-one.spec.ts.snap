exports[`test creates the expected data 1`] = `
Object {
  "aNumber": 123,
  "aString": "abc"
}
`;

exports[`test creates the expected options 1`] = `
Object {
  "computed": Object {
    "aComputedNumber": Object {
      "get": [Function get],
      "set": [Function set]
    },
    "aComputedNumberGetter": [Function get],
    "aComputedString": Object {
      "get": [Function get],
      "set": [Function set]
    },
    "aComputedStringGetter": [Function get]
  },
  "created": [Function anonymous],
  "data": [Function anonymous],
  "methods": Object {
    "aMethod": [Function anonymous],
    "aStringWatch": [Function anonymous]
  },
  "props": Object {
    "aNumberProp": Object {},
    "aNumberPropWithValue": Object {
      "default": 123,
      "required": true,
      "type": [Function Number]
    },
    "aStringProp": Object {},
    "aStringPropWithValue": Object {
      "default": "abc",
      "required": true,
      "type": [Function String]
    }
  },
  "watch": Object {
    "aString": Object {
      "handler": [Function anonymous]
    }
  }
}
`;
